### PR TITLE
SQFP16 Range Validation for Faiss IVF Models

### PR DESCRIPTION
**Pull request by naveentatikonda**
### Description
Faiss SQfp16 Range Validation for Faiss IVF Models and to clip out of range values to fp16 range when the `clip` encoder parameter is set to `true`. 
 
### Issues Resolved
https://github.com/opensearch-project/k-NN/issues/1138
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by ryanbogan — APPROVED** [review]
LGTM!

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java [hunk 1]
```diff
@@ -571,34 +583,30 @@ protected void parseCreateField(ParseContext context, int dimension, SpaceType s
...
-        Map<String, Object> methodComponentParams = knnMethodContext.getMethodComponentContext().getParameters();
+        Map<String, Object> methodComponentParams = methodComponentContext.getParameters();
 
         // The method component parameters should have an encoder
         if (!methodComponentParams.containsKey(METHOD_ENCODER_PARAMETER)) {
             return false;
         }
 
-        MethodComponentContext methodComponentContext = (MethodComponentContext) methodComponentParams.get(METHOD_ENCODER_PARAMETER);
+        MethodComponentContext encoderMethodComponentContext = (MethodComponentContext) methodComponentParams.get(METHOD_ENCODER_PARAMETER);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:602]: Can we check type before casting?

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java [hunk 2]
```diff
@@ -571,34 +583,30 @@ protected void parseCreateField(ParseContext context, int dimension, SpaceType s
...
 
         // returns true if encoder name is "sq" and type is "fp16"
-        return ENCODER_SQ.equals(methodComponentContext.getName())
-            && FAISS_SQ_ENCODER_FP16.equals(methodComponentContext.getParameters().getOrDefault(FAISS_SQ_TYPE, FAISS_SQ_ENCODER_FP16));
+        return ENCODER_SQ.equals(encoderMethodComponentContext.getName())
+            && FAISS_SQ_ENCODER_FP16.equals(
+                encoderMethodComponentContext.getParameters().getOrDefault(FAISS_SQ_TYPE, FAISS_SQ_ENCODER_FP16)
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:607]: Can we just match strictly or can the encoder type for sq not be set?

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java [hunk 3]
```diff
@@ -659,21 +667,19 @@ Optional<byte[]> getBytesFromContext(ParseContext context, int dimension) throws
...
-        boolean isFaissSQfp16Flag = isFaissSQfp16(fieldType().getKnnMethodContext());
+        boolean isFaissSQfp16Flag = isFaissSQfp16(methodComponentContext);
         boolean clipVectorValueToFP16RangeFlag = false;
         if (isFaissSQfp16Flag) {
             clipVectorValueToFP16RangeFlag = isFaissSQClipToFP16RangeEnabled(
-                (MethodComponentContext) fieldType().getKnnMethodContext()
-                    .getMethodComponentContext()
-                    .getParameters()
-                    .get(METHOD_ENCODER_PARAMETER)
+                (MethodComponentContext) methodComponentContext.getParameters().get(METHOD_ENCODER_PARAMETER)
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:682]: Can we cast safely? Check type first

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java (hunk 1 above)
Comment (naveentatikonda) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:607]: updated

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java (hunk 2 above)
Comment (naveentatikonda) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:612]: Yeah, right now as we are only supporting `fp16` for `sq` encoder it is set as default.

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java (hunk 3 above)
Comment (naveentatikonda) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:687]: Here, we are casting it only when the if condition is true which means it is valid as we have already verified above. So, I feel we can skip the validation here.

**Review by jmazanec15 — APPROVED** [review]

**Pull request merged**